### DISCUSSION TITLE
Add where Subquery using the model query builder

### DISCUSSION THREAD

**Discussion by eusonlito** [laravel/framework#46774] [Q&A]
I have a Message Model with a custom Query Builder:

```php
<?php declare(strict_types=1);

namespace App\Domains\Message\Model;

use App\Domains\Message\Model\Builder\Message as Builder;
use Illuminate\Database\Eloquent\Model;

class Message extends Model
{
    /**
     * @var string
     */
    protected $table = 'message';

    /**
     * @param \Illuminate\Database\Query\Builder $query
     *
     * @return \App\Domains\Message\Model\Builder\Message
     */
    public function newEloquentBuilder($query): Builder
    {
        return new Builder($query);
    }
}
```

The Model Query Builder is:

```php
<?php declare(strict_types=1);

namespace App\Domains\Message\Model\Builder;

use App\Domains\Message\Model\Message as Model;
use Illuminate\Database\Eloquent\Builder;

class Message extends Builder
{
    public function byUuid(string $uuid): self
    {
        return $this->where('uuid', $uuid);
    }

    public function byUuidBefore(string $uuid): self
    {
        return $this->where('id', '<', static fn ($q) => $q->select('id')->byUuid($uuid)->limit(1));
    }
}
```

This method returns the error:

```
Call to undefined method Illuminate\\Database\\Query\\Builder::byUuid()
```

The problem is the third parameter in `where`, because is filled with the default Database Builder, instead the current Model Builder. 

Following the framework code, I get:

The `where` is moved from Eloquent to default Query builder: https://github.com/laravel/framework/blob/81c61bdd234150667aaea2028914564317fed32f/src/Illuminate/Database/Eloquent/Builder.php#L290-L301

Now the Query Builder is the default: https://github.com/laravel/framework/blob/81c61bdd234150667aaea2028914564317fed32f/src/Illuminate/Database/Query/Builder.php#L794-L799

And here where the Callback is called: https://github.com/laravel/framework/blob/81c61bdd234150667aaea2028914564317fed32f/src/Illuminate/Database/Query/Builder.php#L1660-L1663

The only way that I have to fix this is creating the query using the Database Builder:

```php
    /**
     * @param string $uuid
     *
     * @return self
     */
    public function byUuidBefore(string $uuid): self
    {
        return $this->where('id', '<', fn ($q) => $this->selectIdByLastUuid($q, $uuid));
    }

    /**
     * @param \Illuminate\Database\Query\Builder $q
     * @param string $uuid
     *
     * @return \Illuminate\Database\Query\Builder
     */
    public function selectIdByLastUuid($q, string $uuid): Builder
    {
        return $q->select('id')
            ->from('message')
            ->where('uuid', $uuid)
            ->orderBy('id', 'DESC')
            ->limit(1);
    }
```

There are some more "Elegant" solution?

Thanks!
Lito.

**Comment by cosmastech** [comment]
Could you put up a public repo with just enough code to display this issue?

**Reply by eusonlito** [reply]
ASAP :)

**Comment by Angel5a** [comment]
Why not to use [dynamic scopes](https://laravel.com/docs/10.x/eloquent#dynamic-scopes)?

**Reply by eusonlito** [reply]
Clean code, models are related with data, query builders should be in a different layer, query builders are IDE friendly by default, models a lot of thin, no `scope` magic, beacuse a lot of reasons 😅
